lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2070,6 +2070,18 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
     }
   }
 
+  /// diagnoseSelfTypedParameters()
+  /// Diagnose instances of (Dynamic)Self as the type of a function arugument.
+  void diagnoseSelfTypedParameters(Type Ty, SourceLoc Loc) {
```

Comment:
Self in a return type where its not at the top level or one level of optional is not allowed either.

```
auto resultTy = func->getResultInterfaceType();
if (resultTy->hasDynamicSelf()) {
  if (!resultTy->lookThroughAllOptionalTypes()->is<DynamicSelfType>()) {
    // we have a dynamic self but its in the wrong place; diagnose
  }
}
```

Reply:
I think the code is there and just needs to be tuned. Which of the following is valid:
```Swift
class C {
    var prop: () -> Self
    var prop: (() -> Self) -> ()
}